### DISCUSSION TITLE
Cloudwatch can't send alarms to SNS topic

### DISCUSSION THREAD

**Discussion by umm0n** [gruntwork-io/knowledge-base#238] [Help] [answered]
Hello everyone! We receive the following error in Cloudwatch whenever an alarm triggers:

` Failed to execute action arn:aws:sns:*****:***********:************-sns-cloudwatch-alarms. Received error: "CloudWatch Alarms does not have authorization to access the SNS topic encryption key." `

We're using the SNS topic module as configured by default with the ref arch, which seems to encrypt the SNS topic with the default aws key. Is this not the correct key to use? What are we missing?

[r:terraform-aws-messaging](https://github.com/gruntwork-io/terraform-aws-messaging)

**Comment by anouarchattouna** [comment, marked as answer]
Hello,

+ If the **default AWS Key Management Service (KMS) key "alias/aws/sns"** is used for SNS topic encryption, then CloudWatch alarms **can't publish messages to the SNS topic**. The key policy of the default AWS KMS key for SNS doesn't allow CloudWatch alarms to perform **_"kms:Decrypt"_** and **_"kms:GenerateDataKey"_** API calls. Because this key is AWS managed, you can't manually edit the policy.
+ You can use a [customer managed key](https://docs.aws.amazon.com/kms/latest/developerguide/concepts.html#customer-cmk) that includes the following permissions under the Statement section of the key policy in order to allow the CloudWatch alarms to publish messages to encrypted SNS topics:
```hcl
{
    "Sid": "Allow_CloudWatch_for_CMK",
    "Effect": "Allow",
    "Principal": {
        "Service":[
            "cloudwatch.amazonaws.com"
        ]
    },
    "Action": [
        "kms:Decrypt",
         "kms:GenerateDataKey*"
    ],
    # better use your SNS topic ARN instead of "*"
    "Resource": "*"
}
```

Best,

**Comment by umm0n** [comment]
Hey @anouarchattouna , thanks for the answer. Adding this CMK indeed fixes the problem and we're able to push to SNS from Cloudwatch. A couple of observations though:
- The ref arch is being shipped (or at least, it was shipped to us) with this default aws kms key configured for the one and only sns topic.
- The sample services and whatnot shipped with the ref arch point to this sns topic via `alarm_sns_topic_arns = [dependency.sns.outputs.topic_arn]`
-  Since Cloudwatch is unable to push to sns without this additional CMK, all modules configured with cloudwatch alarms are effectively "broken" out of the box in the ref arch, at least when it comes to alerting.

My suggestions would be to either:
- Ship the ref arch with a pre-made CMK attached to the sns topic, with the necessary permissions.
- Indicate that this behavior is expected and how to correct it in the onboarding docs for the ref arch.

Again, thanks for your quick response. Cheers!

**Reply by yorinasub17** [reply]
Hi @umm0n , this is indeed a bug with the Ref Arch, and something we need to address. We have created an issue in our tracker and will be prioritizing it accordingly!

**Reply by umm0n** [reply]
Good to know, thanks Yori!
